### PR TITLE
[FIX] mass_mailing: allow duplicate in test mode

### PR DESCRIPTION
**Pull request by yosa-odoo**
Steps to reproduce:
- In Marketing Automation - Campaigns
- Select the demo campaign
- Click on "Launch a test"
- Select Brandon
- Send the two mails from the campaign
- Redo all the steps with a new Test but the same contact

Issue:
- The mails won't be sent

Cause:
- Since for this campaign we already sent these mails to Brandon, the mails will be considered as duplicate in:
https://github.com/odoo/odoo/blob/11b7d89114178fa43545fb8e6224cb4ccbcf8ec1/addons/mail/wizard/mail_compose_message.py#L499

Solution:
We want to make sure that:
- when doing a test, since there is only one linked partner to it, we can send as many mails as we launch tests.
-> repeated tests make sense as we want to fine tune campaigns and maybe get feedback from actual customers
- when launching the actual campaign, there are no duplicates (normal flow) but that eventual test-customers receive the final campaign

In order to do that, we have to filter out from the seen_list
https://github.com/odoo/odoo/blob/9b25a4b5a782146e8c0b41eb75ceea7ac3cd3abf/addons/mass_mailing/models/mailing.py#L736
the test-records.

opw-2810298

### CODE REVIEW

**Comment by yosa-odoo** [comment]
https://github.com/odoo/enterprise/pull/26624

**Comment by yosa-odoo** [comment]
Hello @odoo/rd-sm ,
Could you have a check?
https://github.com/odoo/enterprise/pull/26624

**Comment by tde-banana-odoo** [comment]
Hereby delegating to @std-odoo  as I am helping a bit knowledge team for the next few days :)

File: addons/mass_mailing/models/mailing.py [hunk 1]
```diff
@@ -744,7 +744,9 @@ def _get_seen_list(self):
             SELECT lower(substring(t.%(mail_field)s, '([^ ,;<@]+@[^> ,;]+)'))
               FROM mailing_trace s
```
Comment (std-odoo) [inline addons/mass_mailing/models/mailing.py:745]: i would not include URL in commit message

Those line can be moved if we do a bugfix in it

File: addons/mass_mailing/models/mailing.py [hunk 2]
```diff
@@ -744,7 +744,9 @@ def _get_seen_list(self):
...
               JOIN %(target)s t ON (s.res_id = t.id)
+              %(join_marketing_automation)s
              WHERE substring(t.%(mail_field)s, '([^ ,;<@]+@[^> ,;]+)') IS NOT NULL
+              %(where_marketing_automation)s
```
Comment (std-odoo) [inline addons/mass_mailing/models/mailing.py:749]: weird to have "marketing automation" string in mass mailing, can you rename it `join_domain` or something like that ?

**Review by std-odoo — COMMENTED** [review]
Hi :) 

I just have a few linting remarks, otherwise I don't see a better fix than this one :) 

Shouldn't you fix that in v14 ?

File: addons/mass_mailing/models/mailing.py (hunk 1 above)
Comment (yosa-odoo) [reply addons/mass_mailing/models/mailing.py:745]: Hello, thanks for the review. 

I'm not sure to understand. Are you only talking about the commit message? If yes, indeed in this case, the url is useless. I generally use it with the hash reference in it. I will change that.

Comment (std-odoo) [reply addons/mass_mailing/models/mailing.py:745]: yes about the commit :) thanks :)

**Comment by yosa-odoo** [comment]
Hello for v14, the code is different:
https://github.com/odoo/odoo/blob/15.0/addons/mail/wizard/mail_compose_message.py#L499-L501 -> is not exisiting in 14.0.

I will investigate

**Comment by yosa-odoo** [comment]
@std-odoo, as mentioned the code behaves differently in 14.0. Maybe we can validate this PR so the client is happy and I'll investigate what to do in 14.0.

File: addons/mass_mailing/models/mailing.py [hunk 3]
```diff
@@ -744,7 +744,9 @@ def _get_seen_list(self):
             SELECT lower(substring(t.%(mail_field)s, '([^ ,;<@]+@[^> ,;]+)'))
               FROM mailing_trace s
               JOIN %(target)s t ON (s.res_id = t.id)
```
Comment (std-odoo) [inline addons/mass_mailing/models/mailing.py:746]: comment not solved, you still have URL in your commit message

**Review by std-odoo — COMMENTED** [review]
@yosa-odoo One small comment about your commit message, otherwise LGTM :)

**Comment by kitan191** [comment]
@robodoo delegate+

**Comment by yosa-odoo** [comment]
@robodoo r+

**Pull request closed without merging**